File: sdks/jvm/src/main/java/com/vouchprotocol/core/Vouch.java
package com.vouchprotocol.core;

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;

/**
 * Java SDK for the Vouch Protocol core (#33).
 *
 * A thin JNA layer over the canonical Rust core's C ABI (the cbindgen-generated
 * {@code vouch_core.h}), so the JVM verifies credentials with the exact same
 * bytes as every other Vouch SDK. Binary values are base64 strings; credentials
 * and proofs are JSON strings.
 *
 * Kotlin users may prefer the generated UniFFI binding (vouch_core.kt) bundled
 * in this module; this Java class needs no Kotlin runtime.
 */
public final class Vouch {

    /** Direct mapping of the C ABI. Functions return a heap C string (freed here). */
    interface Lib extends Library {
        Pointer vouch_version();
        Pointer vouch_canonicalize(String json, PointerByReference err);
        Pointer vouch_generate_ed25519(PointerByReference err);
        Pointer vouch_sign(String cred, String seedB64, String vm, String created, PointerByReference err);
        Pointer vouch_build_proof(String cred, String seedB64, String vm, String created, PointerByReference err);
        Pointer vouch_verify_proof(String cred, String pubB64, PointerByReference err);
        Pointer vouch_verify(String cred, String pubB64, String nowIso, long skew, PointerByReference err);
        Pointer vouch_verify_dual(String cred, String edPubB64, String mldsaPubB64, PointerByReference err);
        Pointer vouch_verify_composite(String cred, String edPubB64, String mldsaPubB64, PointerByReference err);
        Pointer vouch_verify_status(String csJson, String slJson, PointerByReference err);
        Pointer vouch_build_delegation_link(String issuer, String subject, String intentJson, String validFrom, String validUntil, String parentProofValue, PointerByReference err);
        Pointer vouch_verify_chain_time_bound(String chainJson, String nowIso, long skew, PointerByReference err);
        void vouch_string_free(Pointer s);
    }

    private static final Lib LIB = Native.load("vouch_core_uniffi", Lib.class);

    /** Thrown when the core reports an error. */
    public static final class VouchException extends RuntimeException {
        public VouchException(String message) { super(message); }
    }

    private Vouch() {}

    private static String take(Pointer result, PointerByReference err) {
        if (result == null) {
            Pointer e = err.getValue();
            String msg = (e == null) ? "unknown error" : e.getString(0);
            if (e != null) LIB.vouch_string_free(e);
            throw new VouchException(msg);
        }
        String s = result.getString(0);
        LIB.vouch_string_free(result);
        return s;
    }

    private static boolean asBool(String s) { return "true".equals(s); }

    /** Version of the underlying core. */
    public static String version() {
        Pointer p = LIB.vouch_version();
        String s = p.getString(0);
        LIB.vouch_string_free(p);
        return s;
    }

    /** RFC 8785 canonicalization of a JSON string. */
    public static String canonicalize(String json) {
        PointerByReference err = new PointerByReference();
        return take(LIB.vouch_canonicalize(json, err), err);
    }

    /** Generate an Ed25519 key pair as JSON {seed_b64, public_b64, multikey, did_key}. */
    public static String generateEd25519() {
        PointerByReference err = new PointerByReference();
        return take(LIB.vouch_generate_ed25519(err), err);
    }

    /** Sign a credential (eddsa-jcs-2022), returning the credential with a proof attached. */
    public static String sign(String credentialJson, String seedB64, String verificationMethod, String created) {
        PointerByReference err = new PointerByReference();
        return take(LIB.vouch_sign(credentialJson, seedB64, verificationMethod, created, err), err);
    }

    /** Build a detached eddsa-jcs-2022 proof object (JSON). */
    public static String buildProof(String credentialJson, String seedB64, String verificationMethod, String created) {
        PointerByReference err = new PointerByReference();
        return take(LIB.vouch_build_proof(credentialJson, seedB64, verificationMethod, created, err), err);
    }

    /** Verify an eddsa-jcs-2022 proof. */
    public static boolean verifyProof(String credentialJson, String publicB64) {
        PointerByReference err = new PointerByReference();
        return asBool(take(LIB.vouch_verify_proof(credentialJson, publicB64, err), err));
    }

    /** Verify proof + validity window, as JSON {proofValid, timeValid, valid}. */
    public static String verify(String credentialJson, String publicB64, String nowIso, long clockSkewSeconds) {
        PointerByReference err = new PointerByReference();
        return take(LIB.vouch_verify(credentialJson, publicB64, nowIso, clockSkewSeconds, err), err);
    }

    /** Verify a dual proof (Ed25519 + ML-DSA-44). */
    public static boolean verifyDual(String credentialJson, String ed25519PublicB64, String mldsaPublicB64) {
        PointerByReference err = new PointerByReference();
        return asBool(take(LIB.vouch_verify_dual(credentialJson, ed25519PublicB64, mldsaPublicB64, err), err));
    }

    /** Verify a v1.6.x composite hybrid proof. */
    public static boolean verifyComposite(String credentialJson, String ed25519PublicB64, String mldsaPublicB64) {
        PointerByReference err = new PointerByReference();
        return asBool(take(LIB.vouch_verify_composite(credentialJson, ed25519PublicB64, mldsaPublicB64, err), err));
    }

    /** Verify a credential's revocation status against a BitstringStatusList credential. */
    public static boolean verifyStatus(String credentialStatusJson, String statusListCredentialJson) {
        PointerByReference err = new PointerByReference();
        return asBool(take(LIB.vouch_verify_status(credentialStatusJson, statusListCredentialJson, err), err));
    }

    /**
     * Build a delegation link. Pass null for any of validFrom, validUntil,
     * parentProofValue to omit it. Returns the link as JSON.
     */
    public static String buildDelegationLink(String issuer, String subject, String intentJson,
            String validFrom, String validUntil, String parentProofValue) {
        PointerByReference err = new PointerByReference();
        return take(LIB.vouch_build_delegation_link(issuer, subject, intentJson, validFrom, validUntil, parentProofValue, err), err);
    }

    /** Validate the time-bound rule over a delegation chain (a JSON array of links). */
    public static boolean verifyChainTimeBound(String chainJson, String nowIso, long clockSkewSeconds) {
        PointerByReference err = new PointerByReference();
        return asBool(take(LIB.vouch_verify_chain_time_bound(chainJson, nowIso, clockSkewSeconds, err), err));
    }
}

File: sdks/swift/Sources/VouchCore/Vouch.swift
import Foundation

/// Idiomatic entry point for the Vouch Protocol core on Apple platforms.
///
/// The `VouchCore` module also exports the lower-level UniFFI functions directly
/// (`canonicalize(json:)`, `generateEd25519()`, ...). This `Vouch` namespace
/// gathers them behind a discoverable surface and adds small conveniences. Every
/// call delegates to the canonical Rust core, so output is byte-identical to the
/// TypeScript, Python, Go, JVM, and .NET SDKs.
public enum Vouch {
    /// Library version of the underlying core.
    public static func version() -> String { VouchCore.version() }

    // MARK: JCS

    /// RFC 8785 canonicalization of a JSON string.
    public static func canonicalize(_ json: String) throws -> String {
        try VouchCore.canonicalize(json: json)
    }

    // MARK: Ed25519 / did:key

    public static func generateEd25519() throws -> Ed25519KeyPair {
        try VouchCore.generateEd25519()
    }
    public static func sign(seed: Data, message: Data) throws -> Data {
        try VouchCore.ed25519Sign(seed: seed, message: message)
    }
    public static func verify(publicKey: Data, message: Data, signature: Data) throws -> Bool {
        try VouchCore.ed25519Verify(publicKey: publicKey, message: message, signature: signature)
    }
    public static func didKey(fromEd25519 publicKey: Data) throws -> String {
        try VouchCore.didKeyFromEd25519(publicKey: publicKey)
    }
    public static func ed25519(fromDidKey did: String) throws -> Data {
        try VouchCore.ed25519FromDidKey(did: did)
    }

    // MARK: Data Integrity (eddsa-jcs-2022)

    /// Sign a credential, returning the credential JSON with a `proof` attached.
    public static func sign(
        _ credentialJson: String,
        seed: Data,
        verificationMethod: String,
        created: String
    ) throws -> String {
        try VouchCore.sign(
            credentialJson: credentialJson,
            seed: seed,
            verificationMethod: verificationMethod,
            created: created
        )
    }

    public static func verifyProof(_ credentialJson: String, publicKey: Data) throws -> Bool {
        try VouchCore.verifyProof(credentialJson: credentialJson, publicKey: publicKey)
    }

    /// Verify the proof and the validity window. `clockSkewSeconds` tolerates drift.
    public static func verify(
        _ credentialJson: String,
        publicKey: Data,
        now: String,
        clockSkewSeconds: Int64 = 30
    ) throws -> VerifyResult {
        try VouchCore.verify(
            credentialJson: credentialJson,
            publicKey: publicKey,
            nowIso: now,
            clockSkewSeconds: clockSkewSeconds
        )
    }

    // MARK: Post-quantum

    public static func generateMldsa44() throws -> MlDsaKeyPair { try VouchCore.generateMldsa44() }

    public static func verifyDual(_ credentialJson: String, ed25519PublicKey: Data, mldsaPublicKey: Data) throws -> Bool {
        try VouchCore.verifyDual(credentialJson: credentialJson, ed25519Public: ed25519PublicKey, mldsaPublic: mldsaPublicKey)
    }
    public static func verifyComposite(_ credentialJson: String, ed25519PublicKey: Data, mldsaPublicKey: Data) throws -> Bool {
        try VouchCore.verifyComposite(credentialJson: credentialJson, ed25519Public: ed25519PublicKey, mldsaPublic: mldsaPublicKey)
    }

    // MARK: Revocation

    public static func verifyStatus(_ credentialStatusJson: String, statusListCredentialJson: String) throws -> Bool {
        try VouchCore.verifyStatus(
            credentialStatusJson: credentialStatusJson,
            statusListCredentialJson: statusListCredentialJson
        )
    }

    // MARK: Delegation

    /// Build a delegation link. `validFrom`, `validUntil`, and
    /// `parentProofValue` are optional. Returns the link as JSON.
    public static func buildDelegationLink(
        issuer: String,
        subject: String,
        intentJson: String,
        validFrom: String? = nil,
        validUntil: String? = nil,
        parentProofValue: String? = nil
    ) throws -> String {
        try VouchCore.buildDelegationLink(
            issuer: issuer,
            subject: subject,
            intentJson: intentJson,
            validFrom: validFrom,
            validUntil: validUntil,
            parentProofValue: parentProofValue
        )
    }

    /// Validate the time-bound rule over a delegation chain (a JSON array of links).
    public static func verifyChainTimeBound(
        _ chainJson: String,
        now: String,
        clockSkewSeconds: Int64 = 30
    ) throws -> Bool {
        try VouchCore.verifyChainTimeBound(
            chainJson: chainJson, nowIso: now, clockSkewSeconds: clockSkewSeconds
        )
    }

    // MARK: FROST(Ed25519) threshold signing (RFC 9591)
    //
    // The aggregated signature is a standard Ed25519 signature, verifiable
    // with verify(_:publicKey:message:signature:) like any other; no new
    // proof type. See vouch_core::threshold for the ceremony and why the full
    // private key is never reconstructed.

    public static func thresholdGenerateKey(minSigners: UInt16, maxSigners: UInt16) throws -> String {
        try VouchCore.thresholdGenerateKey(minSigners: minSigners, maxSigners: maxSigners)
    }
    public static func thresholdCommit(_ keyShareJson: String) throws -> String {
        try VouchCore.thresholdCommit(keyShareJson: keyShareJson)
    }
    public static func thresholdSignShare(
        message: Data, keyShareJson: String, noncesB64: String, commitmentsJson: String
    ) throws -> String {
        try VouchCore.thresholdSignShare(
            message: message, keyShareJson: keyShareJson, noncesB64: noncesB64, commitmentsJson: commitmentsJson
        )
    }
    public static func thresholdAggregate(
        message: Data, commitmentsJson: String, sharesJson: String, groupPublicKeyJson: String
    ) throws -> Data {
        try VouchCore.thresholdAggregate(
            message: message, commitmentsJson: commitmentsJson, sharesJson: sharesJson,
            groupPublicKeyJson: groupPublicKeyJson
        )
    }

    // MARK: Root-identity recovery by Shamir secret sharing
    //
    // Distinct from FROST above: the seed IS reconstructed here, deliberately,
    // for cold recovery of a root identity, not for hot signing. See
    // vouch_core::recovery.

    public static func recoverySplitSecret(_ secretB64: String, threshold: UInt16, shares: UInt16) throws -> String {
        try VouchCore.recoverySplitSecret(secretB64: secretB64, threshold: threshold, shares: shares)
    }
    public static func recoveryCombineShares(_ sharesJson: String) throws -> String {
        try VouchCore.recoveryCombineShares(sharesJson: sharesJson)
    }
    public static func recoverySplitIdentity(_ seedB64: String, threshold: UInt16, shares: UInt16) throws -> String {
        try VouchCore.recoverySplitIdentity(seedB64: seedB64, threshold: threshold, shares: shares)
    }
    public static func recoveryRecoverIdentity(_ sharesJson: String, did: String) throws -> String {
        try VouchCore.recoveryRecoverIdentity(sharesJson: sharesJson, did: did)
    }
}
